File: frontend/src/visualizers/bars.ts
import type { DrawFn } from './types'
import { makeGradient, applyGlow, getFreqBars, hexToRgba } from './types'

export const drawBars: DrawFn = ({ ctx, W, H, t, audio, viz }) => {
  const { barCount, barWidth, barGap, roundness, mirror, glow,
          primaryColor, secondaryColor, colorMode, opacity,
          sensitivity, bassBoost, posX, posY, scale } = viz

  const bars = getFreqBars(audio, barCount, t, sensitivity, bassBoost)

  // Total group width scales with the canvas, controlled by `scale`.
  // barWidth and barGap are treated as a fill ratio so they're canvas-independent.
  const groupW   = Math.min(W * 0.99, W * scale)
  const slotW    = groupW / barCount
  // fillRatio: what fraction of each slot is bar vs gap (barGap=0 → solid, high barGap → thin bars)
  const fillRatio = barWidth / (barWidth + Math.max(0.01, barGap))
  const actualBarW = Math.max(1, slotW * fillRatio)

  const cx     = W * (posX / 100)
  const maxH   = H * 0.45 * scale
  const startX = cx - groupW / 2
  const baseY  = H * (posY / 100)

  ctx.save()

  for (let i = 0; i < barCount; i++) {
    const mag  = bars[i]
    const barH = Math.max(2, mag * maxH)
    const x    = startX + i * slotW + (slotW - actualBarW) / 2   // centered in slot

    const grad = colorMode === 'solid'
      ? hexToRgba(primaryColor, opacity)
      : colorMode === 'rainbow'
        ? `hsla(${(i / barCount) * 360},90%,65%,${opacity})`
        : makeGradient(ctx, 0, baseY, 0, baseY - maxH, secondaryColor, primaryColor, opacity)

    ctx.fillStyle = grad
    applyGlow(ctx, primaryColor, glow)

    const r = Math.min(roundness, actualBarW / 2, barH / 2)

    if (mirror) {
      drawRoundedBar(ctx, x, baseY - barH, actualBarW, barH, r, true)
      ctx.globalAlpha = 0.45
      drawRoundedBar(ctx, x, baseY, actualBarW, barH, r, false)
      ctx.globalAlpha = 1
    } else {
      drawRoundedBar(ctx, x, baseY - barH, actualBarW, barH, r, true)
    }
  }

  ctx.restore()
}

function drawRoundedBar(
  ctx: CanvasRenderingContext2D,
  x: number, y: number, w: number, h: number, r: number, roundTop: boolean,
) {
  if (!roundTop || r <= 0) {
    ctx.fillRect(x, y, w, h)
    return
  }
  ctx.beginPath()
  ctx.moveTo(x + r, y)
  ctx.lineTo(x + w - r, y)
  ctx.quadraticCurveTo(x + w, y, x + w, y + r)
  ctx.lineTo(x + w, y + h)
  ctx.lineTo(x, y + h)
  ctx.lineTo(x, y + r)
  ctx.quadraticCurveTo(x, y, x + r, y)
  ctx.closePath()
  ctx.fill()
}

File: backend/app/services/visualizers/bars.py
"""Port of frontend/src/visualizers/bars.ts"""
from __future__ import annotations

from typing import Optional

from PIL import Image, ImageDraw

from app.services.audio_analyzer import AudioFrame
from app.services.visualizers.types import (
    apply_glow, get_freq_bars, hex_to_rgba, hsl_to_rgb,
    lerp_colors, paste_grad_bar,
)


def draw_bars(
    W: int, H: int, t: float,
    audio: Optional[AudioFrame],
    s,          # RenderSettings
) -> Image.Image:
    layer = Image.new("RGBA", (W, H), (0, 0, 0, 0))
    draw  = ImageDraw.Draw(layer)

    bars      = get_freq_bars(audio, s.bar_count, t, s.sensitivity, s.bass_boost)
    group_w   = min(W * 0.99, W * s.scale)
    slot_w    = group_w / s.bar_count
    fill_r    = s.bar_width / (s.bar_width + max(0.01, s.bar_gap))
    actual_bw = max(1.0, slot_w * fill_r)
    cx        = W * (s.pos_x / 100)
    max_h     = H * 0.45 * s.scale
    start_x   = cx - group_w / 2
    base_y    = H * (s.pos_y / 100)
    radius    = s.roundness

    for i, mag in enumerate(bars):
        bar_h = max(2.0, mag * max_h)
        x     = start_x + i * slot_w + (slot_w - actual_bw) / 2
        x0, y0 = round(x), round(base_y - bar_h)
        x1, y1 = round(x + actual_bw), round(base_y)
        if x1 <= x0:
            continue

        if s.color_mode == "solid":
            color = hex_to_rgba(s.primary_color, s.opacity)
            r = min(radius, (x1 - x0) // 2, (y1 - y0) // 2)
            if r > 0:
                draw.rounded_rectangle([x0, y0, x1, y1], radius=r, fill=color)
            else:
                draw.rectangle([x0, y0, x1, y1], fill=color)

        elif s.color_mode == "rainbow":
            rc, gc, bc = hsl_to_rgb((i / s.bar_count) * 360, 0.9, 0.65)
            color = (rc, gc, bc, round(s.opacity * 255))
            r = min(radius, (x1 - x0) // 2, (y1 - y0) // 2)
            if r > 0:
                draw.rounded_rectangle([x0, y0, x1, y1], radius=r, fill=color)
            else:
                draw.rectangle([x0, y0, x1, y1], fill=color)

        else:  # gradient
            # Canvas: makeGradient(0, baseY, 0, baseY-maxH, secondaryColor, primaryColor)
            # → bottom=secondary, top=primary; bar shows partial slice
            t_top = min(1.0, bar_h / max_h) if max_h > 0 else 0.0
            c_top = lerp_colors(s.secondary_color, s.primary_color, t_top)
            c_top_hex = "#{:02x}{:02x}{:02x}".format(*c_top)
            paste_grad_bar(layer, x0, y0, x1, y1, c_top_hex, s.secondary_color, s.opacity, radius)

        # Mirror (faint reflection below base_y)
        if s.mirror:
            my0, my1 = round(base_y), round(base_y + bar_h)
            if my1 > my0:
                mir_color = hex_to_rgba(s.primary_color, s.opacity * 0.45)
                r = min(radius, (x1 - x0) // 2, (my1 - my0) // 2)
                if r > 0:
                    draw.rounded_rectangle([x0, my0, x1, my1], radius=r, fill=mir_color)
                else:
                    draw.rectangle([x0, my0, x1, my1], fill=mir_color)

    if s.glow > 0:
        layer = apply_glow(layer, s.glow)

    return layer
